File: email-manager/backend/app/routers/templates.py
import re
import uuid

from fastapi import APIRouter, Depends, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_user
from app.core.exceptions import forbidden_exception, not_found_exception
from app.database import get_db
from app.models.template import Template
from app.models.user import User
from app.schemas.template import TemplateCreate, TemplateOut, TemplatePreviewRequest, TemplateUpdate

router = APIRouter(prefix="/templates", tags=["templates"])


@router.get("", response_model=list[TemplateOut])
async def list_templates(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Template).where(Template.user_id == current_user.id).order_by(Template.name)
    )
    return result.scalars().all()


@router.post("", response_model=TemplateOut, status_code=status.HTTP_201_CREATED)
async def create_template(
    body: TemplateCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    template = Template(user_id=current_user.id, **body.model_dump())
    db.add(template)
    await db.commit()
    await db.refresh(template)
    return template


@router.get("/{template_id}", response_model=TemplateOut)
async def get_template(
    template_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await _get_template(template_id, current_user.id, db)


@router.put("/{template_id}", response_model=TemplateOut)
async def update_template(
    template_id: uuid.UUID,
    body: TemplateUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    template = await _get_template(template_id, current_user.id, db)
    for field, value in body.model_dump(exclude_unset=True).items():
        setattr(template, field, value)
    await db.commit()
    await db.refresh(template)
    return template


@router.delete("/{template_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_template(
    template_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    template = await _get_template(template_id, current_user.id, db)
    await db.delete(template)
    await db.commit()


@router.post("/preview")
async def preview_template(
    body: TemplatePreviewRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    template = await _get_template(body.template_id, current_user.id, db)
    rendered_html = substitute_variables(template.body_html, body.variables)
    rendered_text = substitute_variables(template.body_text or "", body.variables)
    return {"body_html": rendered_html, "body_text": rendered_text}


def substitute_variables(text: str, variables: dict[str, str]) -> str:
    """Replace {{variable_name}} placeholders with provided values."""
    def replacer(match):
        key = match.group(1).strip()
        return variables.get(key, match.group(0))
    return re.sub(r"\{\{([^}]+)\}\}", replacer, text)


async def _get_template(template_id: uuid.UUID, user_id: uuid.UUID, db: AsyncSession) -> Template:
    result = await db.execute(select(Template).where(Template.id == template_id))
    template = result.scalar_one_or_none()
    if not template:
        raise not_found_exception
    if template.user_id != user_id:
        raise forbidden_exception
    return template

File: email-manager/frontend/src/api/templates.ts
import { apiClient } from "./client";

export interface Template {
  id: string;
  name: string;
  description: string | null;
  subject_line: string | null;
  body_html: string;
  body_text: string | null;
  variables: { key: string; default: string }[];
  tags: string[];
  use_count: number;
  created_at: string;
  updated_at: string;
}

export async function listTemplates(): Promise<Template[]> {
  const res = await apiClient.get<Template[]>("/templates");
  return res.data;
}

export async function createTemplate(data: Partial<Template> & { name: string; body_html: string }): Promise<Template> {
  const res = await apiClient.post<Template>("/templates", data);
  return res.data;
}

export async function updateTemplate(id: string, data: Partial<Template>): Promise<Template> {
  const res = await apiClient.put<Template>(`/templates/${id}`, data);
  return res.data;
}

export async function deleteTemplate(id: string): Promise<void> {
  await apiClient.delete(`/templates/${id}`);
}

export async function previewTemplate(template_id: string, variables: Record<string, string>) {
  const res = await apiClient.post("/templates/preview", { template_id, variables });
  return res.data as { body_html: string; body_text: string };
}
